### PR TITLE
deps: update protobuf to 3.14

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: update protobuf to 3.14
Additional Description:
There is an unfortunate change in 3.14 that changed `go_package` for WKT, which necessitates updating several go dependencies as well:
- protoc-gen-validate to 2020-11-30
- transitive dependencies from protoc-gen-validate
- gazelle to 0.22.2
- rules_go to 0.25.0
- go to 1.15.5

Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by kyessenov** [comment]
Please provide guidance how to fix deps check CI. Not sure what I understand what it means.

File: api/bazel/repository_locations.bzl
Comment (moderation) [inline api/bazel/repository_locations.bzl:18]: Can we bump to the latest commit?
`1bcea29601b5624234a19b3d7f0ebd9e9984f583`
`2062bbe50eddf3c98490339721fb02b5b5cd78f610f163b98bbf95ba7105553f`

File: api/bazel/repository_locations.bzl
Comment (moderation) [inline api/bazel/repository_locations.bzl:21]: `2020-11-30`

**Review by moderation — CHANGES_REQUESTED** [review]
Can we use this opportunity to bump the Go version at https://github.com/envoyproxy/envoy/blob/master/bazel/dependency_imports.bzl#L19 to `1.15.5`? 

Changes to dependencies get flagged for approval by https://github.com/orgs/envoyproxy/teams/dependency-shepherds. Policy at https://github.com/envoyproxy/envoy/blob/master/DEPENDENCY_POLICY.md#dependency-shepherds

File: api/bazel/repository_locations.bzl
Comment (kyessenov) [reply api/bazel/repository_locations.bzl]: Done

File: api/bazel/repository_locations.bzl
Comment (kyessenov) [reply api/bazel/repository_locations.bzl]: Done

**Comment by kyessenov** [comment]
There is something wrong with scripts:
```
tools/code_format/check_format.py fix
ERROR: From ./resolved.bzl
ERROR: ./resolved.bzl:3940: unexpected direct external dependency on protobuf, use //source/common/protobuf instead.
ERROR: check format failed. run 'tools/code_format/check_format.py fix'
```

**Comment by kyessenov** [comment]
What does the following means:
```
Dependency validation failed, please check metadata in bazel/repository_locations.bzl
Observed dataplane core deps {'com_github_lyft_protoc_gen_star', 'com_google_absl', 'io_bazel_rules_go', 'com_github_c_ares_c_ares', 'com_github_gabime_spdlog', 'net_zlib', 'com_google_protobuf', 'org_golang_google_genproto', 'com_github_tencent_rapidjson', 'com_github_cncf_udpa', 'com_github_spf13_afero', 'com_envoyproxy_protoc_gen_validate', 'boringssl', 'com_github_fmtlib_fmt', 'com_github_cyan4973_xxhash', 'com_google_googleapis', 'org_golang_x_text', 'com_github_iancoleman_strcase', 'com_github_gperftools_gperftools', 'boringssl_fips', 'com_github_libevent_libevent', 'com_github_grpc_grpc', 'com_github_google_tcmalloc', 'com_github_nghttp2_nghttp2', 'com_github_circonus_labs_libcircllhist', 'com_github_zlib_ng_zlib_ng', 'opencensus_proto', 'com_github_jbeder_yaml_cpp', 'com_googlesource_code_re2', 'com_github_nodejs_http_parser'} is not covered by "use_category" implied core deps {'com_google_absl', 'io_bazel_rules_go', 'prometheus_metrics_model', 'bazel_skylib', 'com_github_c_ares_c_ares', 'com_github_gabime_spdlog', 'net_zlib', 'com_google_protobuf', 'com_github_tencent_rapidjson', 'com_github_cncf_udpa', 'com_envoyproxy_protoc_gen_validate', 'boringssl', 'com_github_fmtlib_fmt', 'com_github_cyan4973_xxhash', 'com_github_openzipkin_zipkinapi', 'com_google_googleapis', 'com_github_gperftools_gperftools', 'boringssl_fips', 'com_github_libevent_libevent', 'com_github_grpc_grpc', 'com_github_google_tcmalloc', 'com_github_circonus_labs_libcircllhist', 'com_github_nghttp2_nghttp2', 'com_github_zlib_ng_zlib_ng', 'rules_proto', 'opencensus_proto', 'com_github_apache_skywalking_data_collect_protocol', 'com_github_jbeder_yaml_cpp', 'com_googlesource_code_re2', 'com_github_nodejs_http_parser'}: {'com_github_lyft_protoc_gen_star', 'org_golang_x_text', 'com_github_iancoleman_strcase', 'org_golang_google_genproto', 'com_github_spf13_afero'} are missing
```
Is it because go dependency somehow became part of data plane? How do I tag "go_repository"?

**Comment by kyessenov** [comment]
@htuch Do you know how to fix protoxform build error:
```
Traceback (most recent call last):
  File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/2/execroot/envoy/bazel-out/k8-opt-exec-2B5CBBC6/bin/tools/protoxform/protoxform.runfiles/envoy/tools/protoxform/protoxform.py", line 12, in <module>
    from tools.protoxform import migrate
  File "/build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/2/execroot/envoy/bazel-out/k8-opt-exec-2B5CBBC6/bin/tools/protoxform/protoxform.runfiles/envoy/tools/protoxform/migrate.py", line 14, in <module>
    from google.api import annotations_pb2
ModuleNotFoundError: No module named 'google.api'
```
I can't tell what changed to make it fail...

**Comment by kyessenov** [comment]
I tried to debug this and found that
```python
import google
print(google.__path__)
```
shows `bazel-bin/tools/protoxform/protoxform.runfiles/com_google_protobuf/python/google`. Python part is wrong but I'm not sure where it's coming from. I didn't touch python in this PR at all.

**Comment by adisuissa** [comment]
I've looked at this and found that the difference seems to be in:
`bazel-bin/tools/protoxform/protoxform.runfiles/com_google_protobuf/python/google/protobuf/__init__.py`

More specifically, the protobuf 3.13.0 version has:
```
__version__ = '3.13.0'

if __name__ != '__main__':
  try:
    __import__('pkg_resources').declare_namespace(__name__)
  except ImportError:
    __path__ = __import__('pkgutil').extend_path(__path__, __name__)
```
which is missing in protobuf v3.14.0.

This seems to be coming from: https://github.com/protocolbuffers/protobuf/pull/7902

**Comment by kyessenov** [comment]
https://github.com/protocolbuffers/protobuf/blob/v3.14.0/python/google/__init__.py shows it's there?
But I also noticed it's gone from runfiles.

**Comment by adisuissa** [comment]
I think it's removed from the inner one (in https://github.com/protocolbuffers/protobuf/blob/v3.14.0/python/google/protobuf/__init__.py)

**Comment by kyessenov** [comment]
OK, let me try to add a patch...

**Comment by kyessenov** [comment]
Looks like another CI failure:
```
##[error]Error: Failed find: ENOENT: no such file or directory, lstat '/srv/azure-pipelines/_work/1/a/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/go_sdk/test/fixedbugs/issue27836.dir/�foo.go'
```
There's a weird character in the file name.

**Comment by adisuissa** [comment]
This looks very similar to this bug: https://github.com/bazelbuild/rules_go/issues/1880

And it seems that the function that fixed the bug was modified recently:
https://github.com/bazelbuild/rules_go/pull/2729/files#diff-f06d5b9a2d01e5a186f1b0f8df0f7a0e622bf7b5d481a6448c477665c357b722L139-L162

I wonder if a patch in `bazel/rules_go.patch` that reverts it would work in this case, and I'm not sure how to test because I cannot reproduce on my end:
```
diff --git a/go/private/sdk.bzl b/go/private/sdk.bzl
index c148c9e2..c03cd495 100644
--- a/go/private/sdk.bzl
+++ b/go/private/sdk.bzl
@@ -177,11 +177,24 @@ def _remote_sdk(ctx, urls, strip_prefix, sha256):
     if len(urls) == 0:
         fail("no urls specified")
     ctx.report_progress("Downloading and extracting Go toolchain")
-    ctx.download_and_extract(
-        url = urls,
-        stripPrefix = strip_prefix,
-        sha256 = sha256,
-    )
+    if urls[0].endswith(".tar.gz"):
+        if strip_prefix != "go":
+            fail("strip_prefix not supported")
+        ctx.download(
+            url = urls,
+            sha256 = sha256,
+            output = "go_sdk.tar.gz",
+        )
+        res = ctx.execute(["tar", "-xf", "go_sdk.tar.gz", "--strip-components=1"])
+        if res.return_code:
+            fail("error extracting Go SDK:\n" + res.stdout + res.stderr)
+        ctx.execute(["rm", "go_sdk.tar.gz"])
+    else:
+      ctx.download_and_extract(
+          url = urls,
+          stripPrefix = strip_prefix,
+          sha256 = sha256,
+      )

 def _local_sdk(ctx, path):
     for entry in ["src", "pkg", "bin"]:
```

**Comment by kyessenov** [comment]
@adisuissa I'm more inclined to think this is a bug in Azure task. The tests seem to pass for me locally as well.

**Comment by lizan** [comment]
Yeah seems to be a bug of Azure task. Though Windows failure seems real.

File: bazel/protobuf.patch
Comment (htuch) [inline bazel/protobuf.patch:31]: Do we need to maintain a patch or can we get a change landed upstream to avoid this? CC @envoyproxy/dependency-shepherds

File: test/common/upstream/upstream_impl_test.cc
Comment (htuch) [inline test/common/upstream/upstream_impl_test.cc:1932]: I think I'd prefer to make these tests less sensitive to internals and use more globbing to get at the essential.

**Review by htuch — COMMENTED** [review]
Thanks for taking on the upgrade work, we're seriously out-of-date 😱

File: bazel/protobuf.patch
Comment (kyessenov) [reply bazel/protobuf.patch:31]: Honestly, I don't quite understand what's the deal with google dependencies, python, and bazel. If someone who understands it better could pick it up, I'd appreciate it. I only see it at the level that this was there before, and it's gone, and it works with it.

**Comment by kyessenov** [comment]
@lizan What can I do about SEH exception? It seems like a Windows system exception. Do I need to find a Windows machine to debug?

File: test/common/upstream/upstream_impl_test.cc
Comment (kyessenov) [reply test/common/upstream/upstream_impl_test.cc]: Updated.

**Comment by lizan** [comment]
> @lizan What can I do about SEH exception? It seems like a Windows system exception. Do I need to find a Windows machine to debug?

@davinci26 @wrowe @sunjayBhatia any idea?

**Comment by sunjayBhatia** [comment]
> > @lizan What can I do about SEH exception? It seems like a Windows system exception. Do I need to find a Windows machine to debug?
> 
> @davinci26 @wrowe @sunjayBhatia any idea?

we can run the tests locally possibly in debug mode

also try to run the tests with more logging to see where it crashes

slack thread: https://envoyproxy.slack.com/archives/C78HA81DH/p1607381780485500

File: bazel/protobuf.patch
Comment (htuch) [reply bazel/protobuf.patch:31]: I think this is from https://github.com/protocolbuffers/protobuf/commit/66e3562aafce093d30473555248f22b3fdc88aad. @dlj-NaN do you know why that PR would have caused `google.api` imports to start breaking?

Comment (kyessenov) [reply bazel/protobuf.patch:31]: One thing to keep in mind is that I can't update googleapis because some python rules are being removed (I need to find out why). So there's that.

**Comment by kyessenov** [comment]
@lizan I finally got over windows CI! Now it's just the PublishTestResults CI task. Is there a workaround to stop it from trying to access the bad file? It has nothing to do with test results.

File: bazel/protobuf.patch
Comment (htuch) [reply bazel/protobuf.patch:31]: OK. I don't think any of our build experts has a much better grasp of the Python/Bazel/GoogleAPIs internals, I'd probably start from first principles debugging :(

Comment (kyessenov) [reply bazel/protobuf.patch:31]: All I wanted was to update cel-cpp which needs new protobuf. I ended up fixing python, golang, and windows builds... From the bug description it just sounds that how python protobufs are packaged is fundamentally broken since they cross python module boundaries. I don't what's the right thing to do here.

**Comment by lizan** [comment]
> @lizan I finally got over windows CI! Now it's just the PublishTestResults CI task. Is there a workaround to stop it from trying to access the bad file? It has nothing to do with test results.

Thanks, I'll see what workaround will be

File: .azure-pipelines/bazel.yml
Comment (kyessenov) [inline .azure-pipelines/bazel.yml:70]: Whoa, how did you manage to push to my fork branch?

Comment (lizan) [reply .azure-pipelines/bazel.yml:70]: If you checked "allow edits from maintainers": https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by kyessenov** [comment]
@jayconrod This PR attempts to update protoc to 3.14 and rules_go to 0.25.0. I thought it worked together since I got the protos to compiler, but then the test go program fails to build (https://github.com/envoyproxy/envoy/blob/master/api/test/build/BUILD#L18, target @envoy_api_canonical//test/build:go_build_test) with the message:
```
link: package conflict error: github.com/golang/protobuf/ptypes/any: multiple copies of package passed to linker:
        @com_github_golang_protobuf//ptypes/any:any
        @io_bazel_rules_go//proto/wkt:any_go_proto
```

Is there some workaround for this?

**Comment by kyessenov** [comment]
@jayconrod I've tried to replace ptypes with io_bazel_rules_go, but ended up with

```
link: package conflict error: github.com/golang/protobuf/ptypes/timestamp: multiple copies of package passed to linker:
        @io_bazel_rules_go//proto/wkt:timestamp_go_proto
        @com_github_golang_protobuf//ptypes/timestamp:timestamp

bazel query 'somepath(@envoy_api_canonical//test/build:go_build_test, @com_github_golang_protobuf//ptypes/timestamp:timestamp)'
@envoy_api_canonical//test/build:go_build_test
@envoy_api_canonical//envoy/api/v2:pkg_go_proto
@io_bazel_rules_go//proto/wkt:timestamp_go_proto
@io_bazel_rules_go//proto/wkt:timestamp_wrapper
@com_github_golang_protobuf//ptypes/timestamp:go_default_library
@com_github_golang_protobuf//ptypes/timestamp:timestamp
```

At this point, I'm not sure what to do.

**Comment by jayconrod** [comment]
Hi @kyessenov, have you read through [Avoiding conflicts](https://github.com/bazelbuild/rules_go/blob/master/proto/core.rst#avoiding-conflicts) in the rules_go proto documentation? Which strategy are you taking (generating proto code using `go_proto_library` or using pre-generated .pb.go files)?

If you're using `go_proto_library`, the dependency on `timestamp_go_proto` shouldn't be a problem. Look for other dependencies on `@com_github_golang_protobuf//ptypes/timestamp` outside of that wrapper.

**Comment by kyessenov** [comment]
Thanks @jayconrod, I found out it's opencensus repo that has a conflict.

File: bazel/protobuf.patch
Comment (dlj-NaN) [reply bazel/protobuf.patch:31]: protocolbuffers/protobuf@66e3562 gets pretty deep into the semantics of Python imports and different packaging. The interaction addressed by that commit is very tricky, and may be hidden by varying the order of imports.

(I'm guessing the reason this appeared with a protobuf commit was exactly due to order of imports... `google.auth` comes after `google.api` alphabetically, but due to transitive imports, `google.protobuf` is effectively a very early import in many scenarios.)

The upshot is that there shouldn't be any files named `google/__init__.py` in other installed packages. Just as a first guess, this might be caused by installing googleapis/google-auth-library-python as an egg (maybe via `setup.py install`?), but other `google` packages using native namespaces. (This also gets into what Bazel does in its pip rules....)

Anyhow, some fixes might be:

1.  Remove `google/__init__.py` from googleapis/google-auth-library-python. (@busunkim96 to comment on that.)
2.  Pre-install the package with an up-to-date version of `pip`, which _should_ install using the native namespace.
3.  Update Bazel's pip rules to make sure it does (2).

Comment (busunkim96) [reply bazel/protobuf.patch:31]: > Remove google/__init__.py from googleapis/google-auth-library-python. 

Removing `google/__init__.py` from the `google-auth` will probably be done next quarter as we officially drop 2.7.  google-auth is last on the list since it is transitively required by nearly all the cloud client libraries.

Over the past year we've gradually moved `google-cloud-API` packages to using native namespace packages (example: [google-cloud-texttospeech](https://github.com/googleapis/python-texttospeech/tree/master/google/cloud)) while other libraries like `google-api-core`, `google-cloud-core`, and `google-auth` have kept `google/__init__.py` with pkg_resources style namespace packages. I see now that setuptools [does not recommend](https://packaging.python.org/guides/packaging-namespace-packages/#creating-a-namespace-package) this, but we have not seen customers open issues about it. Our docs always point folks `pip install` packages.

> One thing to keep in mind is that I can't update googleapis because some python rules are being removed (I need to find out why). So there's that.

Are you referring to Bazel rules in https://github.com/googleapis/googleapis or something else?

I don't have much context here so I'm not sure it matters, but the google-cloud-* libraries get their copy of many commonly used protos (`google.api`, `google.type`, ...) from  https://github.com/googleapis/python-api-common-protos, published on PyPI as [google-apis-common-protos](https://pypi.org/project/googleapis-common-protos/)

Comment (moderation) [reply bazel/protobuf.patch:31]: Not sure if this is related but I'm trying to land an update to how we process pip dependencies using the latest `rules_python` at https://github.com/envoyproxy/envoy/pull/14329

**Comment by kyessenov** [comment]
@htuch @moderation This is ready for review again. Let me know if you want to resolve python patch here, or in a followup.

File: bazel/dependency_imports.bzl
Comment (moderation) [inline bazel/dependency_imports.bzl:19]: 1.15.6 is out but I can bump this in a follow up

**Comment by htuch** [comment]
@kyessenov I'm good with merging this if you can followup on the Python protobuf patch as a followup.

**Comment by kyessenov** [comment]
@htuch sure, let's open an issue for tracking then. We'll need to include relevant people, since I don't write python with bazel myself and unaware of best practices.

**Review by htuch — APPROVED** [review]

File: bazel/protobuf.patch
Comment (kyessenov) [reply bazel/protobuf.patch:31]: I opened an issue for further discussion https://github.com/envoyproxy/envoy/issues/14412. 
We're using googleapis/googleapis here because some protos are missing from https://github.com/googleapis/python-api-common-protos (e.g. `google.api.expr` for CEL). Does this mean we can't use PIP approach since there's no pre-generated package I am aware for it?

Comment (moderation) [reply bazel/protobuf.patch:31]: pip handling changes with the latest `rules_python` - https://github.com/bazelbuild/rules_python/releases/tag/0.1.0. PR https://github.com/envoyproxy/envoy/pull/14329 that moves us to that handling is blocked due to Windows (/cc @envoyproxy/windows-dev). In theory pip supports installing from Git - https://pip.pypa.io/en/stable/reference/pip_install/#git but no idea if this works for the current deprecated pip method or the new method we are trying to get to.

**Comment by kyessenov** [comment]
Can we merge this? I have cel-cpp update staged which needs this protobuf version.

**Review by htuch — APPROVED** [review]
LGTM, thanks so much for persisting through the obstacle course run!

**Comment by rgs1** [comment]
@kyessenov fyi the rules_go update breaks rules_docker (for us internally, maybe for others too):

https://github.com/bazelbuild/rules_docker/blob/master/repositories/go_repositories.bzl#L34

which calls go_register_toolchains() without passing a version... (this was fine before, it'd would fallback to a default).

**Comment by kyessenov** [comment]
@rgs1 Sorry about that. It's hard to get around breaking changes since only the latest version of rules_go supports 3.14 protoc but the latest version also turns duplicate linker warnings into errors.

**Comment by rgs1** [comment]
All good, I'll send a PR for rules_docker -- just commenting in case anyone else was hitting this.

**Comment by rgs1** [comment]
https://github.com/bazelbuild/rules_docker/pull/1700 is maybe enough.

**Pull request merged**
